accepted/2020/feature-switch.md

```diff
@@ -0,0 +1,119 @@
+# Feature switch
+The functionality available in the .NET libraries is getting bigger and wider with every release. Applications typically don't need or want all of it, which means when an app includes the libraries as part of it, app size is unnecessarily increased. With technologies like the linker, we have the ability to remove parts of the framework which the app doesn't use. These technologies have limitations based on existing public API patterns and code behavior, sometimes they can't determine what the app really needs. This proposal adds a way to explicitly disable functionality in the framework.
+
+## Goals
+* The ability for developers to declaratively remove functionality in the framework, 3rd party libraries or final app
+* Consistent behavior across different build configurations - the app should get the exact same behavior whether the linker was used on it or not
+* Ability to set different defaults based on the type of application (Blazor, Xamarin, …)
+
+
+## Expected usage
+### Size
+By disabling functionality which is otherwise always available the linker can trim more code from the app reducing its size.
+For example
+* Support for compiled regular expressions. The constructor for regular expression [`Regex(string, RegexOptions)`](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex.-ctor?view=netcore-3.1#System_Text_RegularExpressions_Regex__ctor_System_String_System_Text_RegularExpressions_RegexOptions_) uses the second parameter to determine if it should pre-compile the regular expression or interpret it each time. The code to pre-compile the expression is relatively large and if the app doesn't use this feature it can be removed. But linker typically won't be able to determine that since it's a dynamic behavior. (This is already [conditionally enabled](https://github.com/dotnet/runtime/blob/a83598dd151628672b1d2ff72310cc8c2065a019/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs#L77-L81) only when dynamic code compilation is supported by the runtime.)
+* Security algorithms which are not in use by the app. For example [`AsymmetricalAlgorithm.Create(string)`](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.asymmetricalgorithm.create?view=netcore-3.1#System_Security_Cryptography_AsymmetricAlgorithm_Create_System_String_) can create a large list of algorithms (another factory API).
+
+### Security
+Disable functionality which is not recommended anymore for security reasons. The app may choose to disable such functionality and thus guarantee that nothing in the app can use it.
+For example:
+* `BinaryFormatter`
+* Some cryptography algorithms which are not considered secure anymore. For example [`HashAlgorithm.Create(string)`](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.hashalgorithm.create?view=netcore-3.1#System_Security_Cryptography_HashAlgorithm_Create_System_String_) can create SHA1
```

Comment:
The crypto algorithm thing keeps coming but I don't think that's in the same boat as `BinaryFormatter`. SHA1 and MD5 might not be cryptographically secure, but many document formats and wire protocols require it. For example, if you want to read a Git repo, you need to use SHA1. I don't see us ever disabling those by default. From a linker stand point we probably don't save much either.

Reply:
You can always do a direct `SHA1.Create()` or `MD5.Create()` if you really want these outdated algorithms. This config won't affect that.

Reply:
For example, SHA1 is required as part of the websocket handshake, and ClientWebSocket uses SHA1.Create directly.

Reply:
> From a linker stand point we probably don't save much either.

Every byte saved... is every second (perf) gain!!! 😉